### PR TITLE
Mock services

### PR DESCRIPTION
**Pull request by GordonWuCn**
Add a mock service and an auto-parser to deploy mock services in any kind of DAG. Instruction to deploy is in the `README`.

### CODE REVIEW

File: mock-services/DAG-parser.py [hunk 1]
```diff
@@ -0,0 +1,337 @@
...
+        for i, s in enumerate(services):
+            if i != 0:
+                string += " && "
+            string += f"kubectl create configmap {s}-configmap --from-file {s}/config.yaml"
+
+        self.template = \
+f"""service_account_key_path = "/Users/gordonwu/Downloads/ordinal-motif-270122-b60019637135.json"
+password = "TqzreFzPqXWKdDWQVfsj6Uv3ZX"
+bookinfo_apps_path = "test.yaml"
+add_configmap = "{string}"
```
Comment (AkshatM) [inline mock-services/DAG-parser.py]: This is the only part of `TerraformTemplate` that's changing. The rest are all sensitive values or wouldn't work on my machine. 

Can you place make it so that `TerraformTemplate` only returns this string, and if we need to add it to `values.tfvars`, we can append if it does not exist?

File: mock-services/DAG-parser.py [hunk 2]
```diff
@@ -0,0 +1,337 @@
...
+                            "labels": {
+                                "app": name,
+                                "version": "v1"
+                            }
+                        },
+                        "spec": {
+                            "containers": [
+                                {
+                                    "name": name,
+                                    "image": "docker.io/gordonwu/fake-service:latest" if not bad \
```
Comment (AkshatM) [inline mock-services/DAG-parser.py]: Can you move images to the org?

File: mock-services/DAG-parser.py [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+                            ]
+                        }
+                    }
+                }
+            }
+        if middle:
+            self.template["spec"]["template"]["spec"]["containers"].append(
+                {
+                    "name": "qbox",
+                    "image": "docker.io/gordonwu/qbox:latest",
```
Comment (AkshatM) [inline mock-services/DAG-parser.py]: Use the org image here, please

File: mock-services/DAG-parser.py [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+    f.close()
+    f = open(os.path.join(dir, "test", "stop.sh"), "w")
+    f.write(t.get_stop_script())
+    f.close()
+    st = os.stat(os.path.join(dir, "test", "run.sh"))
+    os.chmod(os.path.join(dir, "test", "run.sh"), st.st_mode | stat.S_IEXEC)
+    st = os.stat(os.path.join(dir, "test", "stop.sh"))
+    os.chmod(os.path.join(dir, "test", "stop.sh"), st.st_mode | stat.S_IEXEC)
+
+    f = open(os.path.join(dir, "test", "values.tfvars"), "w")
```
Comment (AkshatM) [inline mock-services/DAG-parser.py]: Rather than hardcode the file contents completely, can you parse the file as `=`-delimited lines, find the `add_configmap` key, and update it before writing to the file? This prevents sensitive values and machine-specific values being written.

File: mock-services/fake-service/service.js [hunk 5]
```diff
@@ -0,0 +1,113 @@
+// Copyright 2017 Istio Authors
```
Comment (AkshatM) [inline mock-services/fake-service/service.js]: See how to [make this work with RabbitMQ here](https://github.com/Brown-CS2952-cwu-amahajan/baseline-qbox-cluster/pull/1/files#diff-64f0e241fe6d385b1dcb8a7cff01c43a). 

tl;dr add a different file (in PR above, it's called `rabbitmq.js`) containing code for a RabbitMQ consumer and run it as a [forked child process](https://nodejs.org/api/child_process.html#child_process_child_process_fork_modulepath_args_options) in `service.js`. You can use interprocess communication (IPC) to communicate between the parent and child - here is an example on [how to do it](https://nodejs.org/api/child_process.html#child_process_subprocess_send_message_sendhandle_options_callback)

you need to do it this way because RabbitMQ client is blocking and polling, meaning you can't run a webserver _and_ RabbitMQ consumer code in the same process.

**Review by AkshatM — CHANGES_REQUESTED** [review]
1. Please move images or update references of images to org repo
2. Please update handling of Terraform `value.tfvars` so we're not hardcoding sensitivie values or machine-specific values. 
3. Please add RabbitMQ bindings - I have provided an example for how to make it work with `dispatcher`. 

Q: Is there a reason we are duplicating all Terraform code here? You can update the existing code in the parent directory if necessary. If you clone `terraform-backend.tf`, the state file will get overwritten if someone is running the code in the parent directory. 

For your work, it looks like all you really need to do is update the`value.tfvars` file, and even then you only need to loop over configmaps. Terraform provides [looping constructs](https://www.hashicorp.com/blog/hashicorp-terraform-0-12-preview-for-and-for-each/), so you can use those.

File: mock-services/fake-service/service.js (hunk 5 above)
Comment (AkshatM) [reply mock-services/fake-service/service.js]: I will be taking care of this, so resolving convo.

File: mock-services/lib/populate-cluster.tf [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+
+resource "null_resource" "add_configmap" {
+  provisioner "local-exec" {
+    command = var.add_configmap
+    interpreter = ["/bin/bash", "-c"]
+  }
+  depends_on = [google_container_cluster.primary]
+}
+
+resource "kubectl_manifest" "bookinfo" {
```
Comment (AkshatM) [inline mock-services/lib/populate-cluster.tf]: You can delete this and the `istio-microservices` folder now. We have no use for either.

**Review by AkshatM — APPROVED** [review]
